raft.go

```diff
@@ -1692,29 +1704,38 @@ type voteResult struct {
 // ourself. This has the side affecting of incrementing the current term. The
 // response channel returned is used to wait for all the responses (including a
 // vote for ourself). This must only be called from the main thread.
-func (r *Raft) electSelf() <-chan *voteResult {
+func (r *Raft) electSelf(preVote bool) <-chan *voteResult {
 	// Create a response channel
 	respCh := make(chan *voteResult, len(r.configurations.latest.Servers))
 
+	// PreVote campaigns are sent on the next term.
+	term := r.getCurrentTerm() + 1
+	if preVote {
```

Comment:
I'm confused about this part. We increment the current term by two if preVote is true; I would have thought we would leave it unchanged. 

Part of my confusion I'm sure is because I've not found a very detailed description of Pre-Vote in the literature. 

In Ongaro's thesis Pre-Vote is described as 'In the Pre-Vote algorithm, a candidate only increments its term if it first learns from a majority of the cluster that they would be willing to grant the candidate their votes' (9.6, pg 137 in my copy). Elsewhere (4.2.3 disruptive servers) "Only if the candidate believed it could get votes from a majority of the cluster would it increment its term and start a normal election."

Looking at the algorithm description in "4 modifications for Raft consensus" I *think* the intention is to increment by one, but they also talk about the need to avoid 'term inflation' and how this is avoided by Pre-Vote, which I freely admit I don't fully understand how to reconcile.

I'm still looking at other implementations for insight.

Reply:
> Part of my confusion I'm sure is because I've not found a very detailed description of Pre-Vote in the literature.

This took some effort to actually originally page this in as not to disrupt the original implementation, I'll have to look over my old notes about this.

Sorry, I can't be much more helpful atm.

Reply:
Hey @SimonRichardson, @markan,
my understanding based on the same literature you are reading @markan is that a node should not increment it's own term for pre-vote to avoid term inflation but still need to send a pre-vote request with a term equal to what it would send if asking for a normal vote.
Based on this, I think the correct way of dealing with this in here is to only `setCurrentTerm` if `preVote` is false and only increment the term once. That way the node will send the exact same term for pre-vote and for normal vote if it decide to start an election after the pre-vote.
pre-vote should not increment the node term.

Reply:
> Based on this, I think the correct way of dealing with this in here is to only setCurrentTerm if preVote is false and only increment the term once. That way the node will send the exact same term for pre-vote and for normal vote if it decide to start an election after the pre-vote.

It seems like this is correct. I've spent some time reading back through this. Originally my notes say that it should be sent on the next term, but reading back through the papers we should send this on the same term as a normal vote.